src/lib/pickles/wrap_verifier.ml

```diff
@@ -432,33 +444,45 @@ struct
     end
 
     let combine batch ~xi without_bound with_bound =
+      let reduce_point p =
+        let point = ref (Point.underlying p.(Array.length p - 1)) in
+        for i = Array.length p - 2 downto 0 do
+          point := Point.add p.(i) (Scalar_challenge.endo !point xi)
+        done ;
+        !point
+      in
       let { Curve_opt.non_zero; point } =
         Pcs_batch.combine_split_commitments batch
+          ~reduce_with_degree_bound:(fun _ -> assert false)
+          ~reduce_without_degree_bound:(fun x -> [ x ])
           ~scale_and_add:(fun ~(acc : Curve_opt.t) ~xi
-                              (p : (Point.t, Boolean.var) Opt.t) ->
+                              (p : (Point.t array, Boolean.var) Opt.t) ->
             (* match acc.non_zero, keep with
                | false, false -> acc
                | true, false -> acc
                | false, true -> { point= p; non_zero= true }
                | true, true -> { point= p + xi * acc; non_zero= true }
             *)
             let point keep p =
-              let point =
+              let base_point =
+                let p = p.(Array.length p - 1) in
                 Inner_curve.(
-                  if_ keep
-                    ~then_:
-                      (if_ acc.non_zero
-                         ~then_:
-                           (Point.add p (Scalar_challenge.endo acc.point xi))
-                         ~else_:
-                           ((* In this branch, the accumulator was zero, so there is no harm in
-                               putting the potentially junk underlying point here. *)
-                            Point.underlying p ) )
-                    ~else_:acc.point)
+                  if_ acc.non_zero
+                    ~then_:(Point.add p (Scalar_challenge.endo acc.point xi))
+                    ~else_:
+                      ((* In this branch, the accumulator was zero, so there is no harm in
+                          putting the potentially junk underlying point here. *)
+                       Point.underlying p ))
               in
-              let non_zero =
-                Boolean.(keep &&& Point.finite p ||| acc.non_zero)
+              let point = ref base_point in
+              for i = Array.length p - 2 downto 0 do
```

Comment:
That's going to give me nightmares tonight this imperative code. Hope I'll be better on Monday :)